sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager+TransferDelegate.swift

```diff
@@ -50,17 +50,43 @@ extension URLSessionTransferManager: ResumableOperationQueueDelegate {
         default:
             if let blobTransfer = transfer as? BlobTransfer {
                 let progress = blobTransfer.progress
+
+                // avoid saving the context or sending multiple messages when the progress has not actually changed
                 if blobTransfer.currentProgress < progress {
                     blobTransfer.currentProgress = progress
                     delegate?.transfer(transfer, didUpdateWithState: state, andProgress: progress)
+                } else {
+                    return
                 }
             } else {
                 delegate?.transfer(transfer, didUpdateWithState: state)
             }
         }
+        saveContext()
+    }
+
+    func transfersDidUpdate(_ transfers: [Transfer]) {
+        guard let restorationId = (transfers.first as? TransferImpl)?.clientRestorationId else { return }
+        guard let delegate = client(forRestorationId: restorationId) as? TransferDelegate else { return }
+        delegate.transfersDidUpdate(transfers)
+        saveContext()
+    }
+
+    func transferDidComplete(_ transfer: Transfer) {
+        guard let restorationId = (transfer as? TransferImpl)?.clientRestorationId else { return }
+        guard let delegate = client(forRestorationId: restorationId) as? TransferDelegate else { return }
+        delegate.transferDidComplete(transfer)
+        saveContext()
     }
 
-    internal func operations(_: [ResumableOperation]?, didChangeState _: TransferState) {
+    func transfer(_ transfer: Transfer, didFailWithError error: Error) {
+        guard let restorationId = (transfer as? TransferImpl)?.clientRestorationId else { return }
+        guard let delegate = client(forRestorationId: restorationId) as? TransferDelegate else { return }
+        delegate.transfer(transfer, didFailWithError: error)
         saveContext()
     }
+
+    func client(forRestorationId restorationId: String) -> PipelineClient? {
```

Comment:
This should probably move to `URLSessionTransferManager.swift` since its use is not exclusive to this protocol extension.

Reply:
I kept it here because it's part of the `TransferDelegate` protocol.

Reply:
Oh, I missed that >< Good call.